dedent/dedent.go

```diff
@@ -1,34 +1,73 @@
 package dedent
 
 import (
-	"strings"
+	"bytes"
 )
 
 // String automatically detects the maximum indentation shared by all lines and
 // trims them accordingly.
 func String(s string) string {
-	lines := strings.Split(s, "\n")
-	minIndent := -1
+	indent := minIndent(s)
+	if indent == 0 {
+		return s
+	}
 
-	for _, l := range lines {
-		if len(l) == 0 {
-			continue
-		}
+	return dedent(s, indent)
+}
 
-		indent := len(l) - len(strings.TrimLeft(l, " "))
-		if minIndent == -1 || indent < minIndent {
-			minIndent = indent
+func minIndent(s string) int {
+	var (
+		curIndent    int
+		minIndent    int
+		shouldAppend = true
+	)
+	for i := 0; i < len(s); i++ {
+		switch s[i] {
+		case ' ':
```

Comment:
I actually quite like this approach for one specific reason: it gives us the chance to not only check against spaces, but also other whitespace characters and tabs. See: `unicode.IsSpace`.

Reply:
Thank for pointing that out! Maybe we should just append '\t' in this situation.

If I understood it correctly, both "  ", " \t", "\t ", "\t\t" are the same?